### PR TITLE
Fix parsing resp3 dicts

### LINKED ISSUE #2981: resp3 parser async parser incorrectly parses maps
Issue discovered in pr #2947

the evaulation order of `key:val` in `{key():val() for foo in bar}` dict comprehension is undefined in Python.  It mostly is left-right,
but some versions, notably recent 3.7 versions, don't behave that way.  It is best to use explicit ordering when reading from  a stream.

This had already been fixed for __non-async__ parsing, but in pr #2947 caused some scheduling behaviour which triggered this to happen. A fix for __async__ has been added there.

### PR DESCRIPTION
**Pull request by kristjanvalur**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Do tests and lints pass with this change?
- [x] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [x] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

Fixes #2981

### CODE REVIEW

File: redis/asyncio/connection.py [hunk 1]
```diff
@@ -493,6 +493,7 @@ async def read_response(
                 and self.protocol in ["3", 3]
                 and not HIREDIS_AVAILABLE
             ):
+                read_timeout = None
```
Comment (dvora-h) [inline redis/asyncio/connection.py:496]: why do you need this?

File: redis/_parsers/resp3.py [hunk 2]
```diff
@@ -225,12 +226,16 @@ async def _read_response(
                 pass
         # map response
         elif byte == b"%":
-            response = {
-                (await self._read_response(disable_decoding=disable_decoding)): (
-                    await self._read_response(disable_decoding=disable_decoding)
+            # we use this approach and not dict comprehension here
+            # because this dict comprehension fails in python 3.7
+            # (evaluation order of key:val is _undefined_ in Python)
```
Comment (dvora-h) [inline redis/_parsers/resp3.py:231]: Do you know if it's _undefined_ also in python 3.8+?

File: redis/asyncio/connection.py (hunk 1 above)
Comment (kristjanvalur) [reply redis/asyncio/connection.py:496]: ah, debugging code which accidentally crept in, good catch!

File: redis/_parsers/resp3.py (hunk 2 above)
Comment (dvora-h) [reply redis/_parsers/resp3.py]: I'm not sure that we need to fix it because python 3.7 is EOL and we are going to drop support on it in 5.1

Comment (kristjanvalur) [reply redis/_parsers/resp3.py]: but it is still supported, though.  as it is, 3.7 will fail parsing resp3.  There is already a fix for the "sync" version, why not fix the async one?

Comment (kristjanvalur) [reply redis/_parsers/resp3.py]: Not pushing for it, though.  just putting in an argument :).   For my other PR where this came up, I can certainly put in a skip for 3.7

**Review by dvora-h — APPROVED** [review]

**Pull request merged**